kubernetes/pods.go

```diff
@@ -0,0 +1,92 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apisv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+const resyncInterval = 1 * time.Minute
+
+// PodInformer is a event handler for Pod events registered to, that builds a local list of valid and relevant pods
+// and sends an event to the endpoint channel, triggering a resync of the targets.
+func (a *Adapter) PodInformer(ctx context.Context, endpointChan chan<- []string) (err error) {
+	podEndpoints := sync.Map{}
+
+	log.Infof("Watching for Pods with labelselector %s in namespace %s", a.cniPodLabelSelector, a.cniPodNamespace)
+	factory := informers.NewSharedInformerFactoryWithOptions(a.clientset, resyncInterval, informers.WithNamespace(a.cniPodNamespace),
+		informers.WithTweakListOptions(func(options *apisv1.ListOptions) { options.LabelSelector = a.cniPodLabelSelector }))
+
+	informer := factory.Core().V1().Pods().Informer()
+	factory.Start(ctx.Done())
```

Comment:
`factory.Start()` apparently [starts](https://github.com/kubernetes/client-go/blob/3255cdcd9cf0ce5ea718c97f4aea519130649aca/informers/factory.go#L127-L138) informers but we configure informer EventHandler later which seems wrong.

From this perspective `informer.Run(ctx.Done())` also feels wrong as factory.Start already runs it.

Reply:
I guess we can do what others do:
workqueue: https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go#L128-L132
fakeclient: https://github.com/kubernetes/client-go/blob/master/examples/fake-client/main_test.go#L69-L72

The cache.WaitForCacheSync() does what we need:
https://github.com/kubernetes/client-go/blob/3255cdcd9cf0ce5ea718c97f4aea519130649aca/informers/factory.go#L141
https://github.com/kubernetes/client-go/blob/3255cdcd9cf0ce5ea718c97f4aea519130649aca/tools/cache/shared_informer.go#L254

I also asked in sig-apimachinery, if they consider this safe or if we have to do something else.

Reply:
so Kubernetes sig-apimachinery says that it's not an issue because List is used before Watch, which means there is no race.

Reply:
@szuecs  
In another discussion https://github.com/zalando-incubator/kube-ingress-aws-controller/pull/461#discussion_r788818501 this issue https://github.com/kubernetes/kubernetes/issues/92777#issuecomment-862559059 was brought up which is exactly about the problem we have now - configuring informer after factory start.

Reply:
Then we should just move the AddEventhandler call with body to here

Reply:
It certainly makes no difference if AddEventhandler is added before or after the start. Eventually I don't care, but I can prove that even via the unit test.
@AlexanderYastrebov you're misreading the issue https://github.com/kubernetes/kubernetes/issues/92777#issuecomment-862559059, it is relevant to the place where WaitForCacheSync is placed regarding informers.Start

Reply:
See the sources https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L552-L556 that explain how a safe join happens to a running informer

Reply:
So this makes sure we get all entries from a fresh LIST, right?

Reply:
Yes- so theoretically we could omit the explicit List before since we get this one implicitly by late binding, but it serves two reasons so I'd keep it

Reply:
> So this makes sure we get all entries from a fresh LIST, right?
> Yes- so theoretically we could omit the explicit List

Idk, I see it uses `addNotification`  but we only handle updates.